Knoblauchpilze/user-service
# feat: Added an e2e tests step

# Work

## Motivation
In order to validate that the docker image works as expected we want to have a very simple set of E2E tests. The purpose of those tests is to verify if with a setup similar to what we have in production we can make some requests to the server and this has the expected results.

## How to achieve this?
We want to leverage the [service containers](https://docs.github.com/en/actions/use-cases-and-examples/using-containerized-services/about-service-containers) offered by github actions to:
* spin-up a `postgresql` database.
* start a docker container that was just built in the CI for the service.
* make a couple of `curl` requests to the server and observe if the responses are matching what we expect.

## Are there obvious issues?
Ideally we would like to create the `user-service` as a service container, just like the database. However there are a number of challenges to overcome for this to work:
* the database is not created when we start the `user-service`.
* the users/passwords are not created yet.
* the database is not migrated.

We could work around some of those limitations: for example the user does not need to be the `user_service_manager` for the service to access the database: any user can do as long as it has the permission on the schema.

The problem is that we create a custom schema which is not available from the start and this prevents to 'just migrate the schema'.

Due to this, we went for a slightly different approach: instead of using a service container for the `user-service` we just pull the latest docker image, then run it and execute the `curl` commands in the same step. It is slightly more cumbersome but it works.

## Additional considerations

In terms or orchestration, we only run the E2E tests if the tests were successful and if the build of the docker image succeeded.

# Tests

Below is a view of how the E2E tests step looks like:

![Screenshot from 2024-11-23 17-04-32](https://github.com/user-attachments/assets/08630ed4-0b10-49e8-88a4-5b6f7ad26ff5)

Important steps are details in the picture. Basically we:
* set up the `postgres` server as a service container
* download the `user-service` image
* run it
* execute `curl` commands
* stop the server
* verify the results of the `curl` commands

## Failure due to bad response envelope

In [f7a0f44 and 4b53689](https://github.com/Knoblauchpilze/user-service/compare/cda16c27f7033bfd22fb8e37868f444f3c504994..4b536895d5327bd816e0055eaa1df87231b60fb7) we voluntarily changed what the response envelope looks like. This (as expected) breaks the E2E tests (see [CI run](https://github.com/Knoblauchpilze/user-service/actions/runs/11988633946/job/33424142535)):

![image](https://github.com/user-attachments/assets/5d4f61d9-1503-4afe-8053-2ba93fb54d81)

## Failure due to bad http status code:

In [00816e5](https://github.com/Knoblauchpilze/user-service/pull/7/commits/00816e579354c9eabe80a548f5daf396281b47b1) we voluntarily changed the status code returned by the list endpoint. This (also as expected) breaks the E2E tests (see [CI run](https://github.com/Knoblauchpilze/user-service/actions/runs/11988671655/job/33424220787)):

![image](https://github.com/user-attachments/assets/8f9113a0-75dd-4aae-99ca-677a7c1cd14a)

# Future work

Currently we only perform minimal testing for the server: the main goal is to make sure that the container is able to communicate with the outside world and able to access the database as needed.

We could refine those tests to include verification of other workflows: for example creating a user (and asserting that it is in the database), getting the user we just created, etc.

This could be done in the future but there are several unknowns:
* which framework to use to perform the testing? Using bash is not really realistic.
* do we want to accumulate the coverage for such tests?

Below are a couple of links that we found while researching the topic:
* Reddit post [E2E tests for go](https://www.reddit.com/r/golang/comments/114wbr2/e2e_tests_for_go/).
* Medium post to [measure E2E tests coverage](https://medium.com/@madworx/measuring-code-coverage-in-end-to-end-test-scenarios-with-golang-3c6c9d8a93ba).

This could be revisited later.
